x/ccv/provider/keeper/msg_server.go

```diff
@@ -309,16 +305,126 @@ func (k msgServer) SetConsumerCommissionRate(goCtx context.Context, msg *types.M
 	return &types.MsgSetConsumerCommissionRateResponse{}, nil
 }
 
-func (k msgServer) ConsumerModification(goCtx context.Context, msg *types.MsgConsumerModification) (*types.MsgConsumerModificationResponse, error) {
-	if k.GetAuthority() != msg.Authority {
-		return nil, errorsmod.Wrapf(types.ErrUnauthorized, "expected %s, got %s", k.GetAuthority(), msg.Authority)
+// CreateConsumer creates a consumer chain
+func (k msgServer) CreateConsumer(goCtx context.Context, msg *types.MsgCreateConsumer) (*types.MsgCreateConsumerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	initGas := ctx.GasMeter().GasConsumed()
+
+	consumerId := k.FetchAndIncrementConsumerId(ctx)
+
+	k.SetConsumerOwnerAddress(ctx, consumerId, msg.Signer)
+	k.SetConsumerChainId(ctx, consumerId, msg.ChainId)
+
+	if err := k.SetConsumerMetadata(ctx, consumerId, msg.Metadata); err != nil {
+		return &types.MsgCreateConsumerResponse{}, err
+	}
+
+	// initialization parameters are optional and hence could be nil
+	if msg.InitializationParameters != nil {
+		if err := k.SetConsumerInitializationParameters(ctx, consumerId, *msg.InitializationParameters); err != nil {
+			return &types.MsgCreateConsumerResponse{}, err
+		}
+	}
+
+	// power-shaping parameters are optional and hence could be null
+	if msg.PowerShapingParameters != nil {
+		if err := k.SetConsumerPowerShapingParameters(ctx, consumerId, *msg.PowerShapingParameters); err != nil {
+			return &types.MsgCreateConsumerResponse{}, err
+		}
 	}
 
+	if k.CanLaunch(ctx, consumerId) {
+		k.SetConsumerPhase(ctx, consumerId, Initialized)
+		k.PrepareConsumerForLaunch(ctx, consumerId, nil, *msg.InitializationParameters.SpawnTime)
+	} else {
+		k.SetConsumerPhase(ctx, consumerId, Registered)
+	}
+
+	gasAfter := ctx.GasMeter().GasConsumed()
+	ctx.GasMeter().ConsumeGas(gasAfter-initGas, "creating a chain has an additional cost during spawn time, "+
+		"so charging double the gas here")
+
+	return &types.MsgCreateConsumerResponse{ConsumerId: consumerId}, nil
+}
+
+// UpdateConsumer updates the record of a consumer chain
+func (k msgServer) UpdateConsumer(goCtx context.Context, msg *types.MsgUpdateConsumer) (*types.MsgUpdateConsumerResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	err := k.Keeper.HandleConsumerModificationProposal(ctx, msg)
+	consumerId := msg.ConsumerId
+
+	ownerAddress, err := k.Keeper.GetConsumerOwnerAddress(ctx, consumerId)
 	if err != nil {
-		return nil, errorsmod.Wrapf(err, "failed handling ConsumerModification proposal")
+		// TODO (PERMISSIONLESS): use a better error message
+		return &types.MsgUpdateConsumerResponse{}, errorsmod.Wrapf(types.ErrInvalidConsumerId, "cannot retrieve owner address %s", ownerAddress)
+	}
+
+	if msg.PowerShapingParameters != nil && msg.PowerShapingParameters.IsTop_NSet && msg.PowerShapingParameters.Top_N > 0 && msg.Signer != k.GetAuthority() {
```

Comment:
This should be part of the update PowerShapingParameters logic, i.e., 
```
if msg.PowerShapingParameters != nil {
    // check if transition from Opt-In to TopN is possible 
    if msg.PowerShapingParameters.IsTop_NSet && msg.PowerShapingParameters.Top_N > 0 && msg.Signer != k.GetAuthority() {
        // error -- this transition needs to go through governance
    }

    // check if transition from TopN to Opt-In is possible
    if msg.PowerShapingParameters.IsTop_NSet && msg.PowerShapingParameters.Top_N == 0 {
        if oldPowerShapingParameters, err := k.Keeper.GetConsumerPowerShapingParameters(ctx, msg.ConsumerId); err == nil {
            if oldPowerShapingParameters.Top_N > 0 {
                if msg.NewOwnerAddress is a valid address {
                    // chain becomes an Opt In chain, hence we change the owner address
                    k.Keeper.SetConsumerOwnerAddress(ctx, consumerId, msg.NewOwnerAddress)
                }
            }
        }
    }

    k.Keeper.SetConsumerInitializationParameters(ctx, msg.ConsumerId, *msg.InitializationParameters)
}
```

Reply:
Before update the PowerShapingParameters, you need to authenticate the message:
```
if msg.Signer == ownerAddress { 
    if ownerAddress != k.GetAuthority() {
        // the gov module shouldn't change ownership of a TopN chain to another address 
        if msg.NewOwnerAddress is a valid address { 
            k.Keeper.SetConsumerOwnerAddress(ctx, consumerId, msg.NewOwnerAddress) 
        }
    }
} else if msg.Signer == k.GetAuthority() {
    if ownerAddress != k.GetAuthority() {
        // in what conditions can the governance take over an opt-in chain? 
    }
} else {
    // error: ErrUnauthorized (expecting either owner address or gov module address) 
}
```
The main question here is in what conditions can the governance take over an opt-in chain? If always, then we just update the consumer's ownership `k.GetAuthority()`.